[move-book] Extending Book by Move 2.0 Documentation (#567)
aptos-labs/developer-docs: apps/nextra/pages/en/build/smart-contracts/book/enums.mdx

```move
enum Color has copy, drop, store, key { Red, Blue, Green }
```

Enum types can also have the `key` ability and appear as roots of data in global storage. A common usage of enums in this context is versioning of data:

```move
enum VersionedData has key {
  V1{name: String}
  V2{name: String, age: u64}
}
```

Similar to structs, enum types can be generic and take positional arguments. For example, the type below represents a generic result type, where the variant constructors use positional instead of named arguments (see also [positional structs](structs-and-resources.mdx#positional-structs)).

```move
enum Result<T> has copy, drop, store {
  Err(u64),
  Ok(T)
}
```

## Constructing Enum Values 

An enum value is constructed similarly to a struct value:

```move
let s: String;
let data = VersionedData::V1{name: s};
```

If the enum variant has no fields, the braces can also be omitted:

```move
let color = Color::Blue;
```

## Name Resolution for Enum Variants 

The variant names for an enum need to be qualified by the enum type name, as in `VersionedData::V1`. 

> > Note:" Aliasing via the `use` clause is currently not supported for enum variants, but will be added in later language versions

Reviewer: ```suggestion
> Note: Aliasing via the `use` clause is currently not supported for enum variants, but will be added in later language versions
```

> # Enums

Reviewer 2: In the linked nextra preview, I would have expected "Enums" to appear in the left bar under "Basic Concepts", perhaps right under "Structs and Resources".

I only see it under "Reference", just below "Coding Conventions".

_Since language version 2.0_

> Enum types are similar to struct types but support defining multiple *variants* of the data layout. Each variant has its distinct set of fields. In expressions, tools for testing, matching, and deconstructing enum variants are supported. 

Reviewer 2: ```suggestion
Enum types are similar to struct types but support defining multiple *variants* of the data layout. Each variant has its distinct set of fields. Enum variants are supported in expressions, tools for testing, matching, and deconstructing. 
```

## Declaration of Enum Types 

An enum type declaration lists the number of different variants, as seen in the example below:

```move
enum Shape {
    Circle{radius: u64},
    Rectangle{width: u64, height: u64}
}
```

There can be zero or more fields for an enum variant. If no arguments are given, the braces can also be omitted, declaring simple values:

```move
enum Color {
  Red, Blue, Green
}
```

Similar to struct types, enum types can have abilities. For example, for the `Color` enum type it is more appropriate to make it copyable, droppable, and storable, similar like primitive number types:

```move

> enum Color has copy, drop, store, key { ... }

Reviewer 2: ```suggestion
enum Color has copy, drop, store, key { Red, Blue, Green }
```

Reviewer 2: Because `..` is a new syntax we are also introducing, having `...` may lead readers to think that is also some new syntax.

```

Enum types can also have the key ability and appear as roots of data in global storage. A common usage of enums in this context is versioning of data:

```move
enum VersionedData has key {
  V1{name: String}
  V2{name: String, age: u64}
}
```

Similar to structs, enum types can be generic. For example, the type below represents a generic result type, where the variant constructors use positional instead of named arguments (see also [positional structs](structs-and-resources.mdx#positional-structs)).

```move
enum Result<T> has copy, drop, store {
  Err(u64),
  Ok(T)
}
```

## Constructing Enum Values 

An enum value is constructed similarly to a struct value:

```move
let s: String;
let data = VersionedData::V1{name: s};
```

If the enum variant has no fields, the braces can also be omitted:

```move
let color = Color::Blue;
```

## Name Resolution for Enum Variants 

An enum variant names need to be qualified by the enum type name, as in `VersionedData::V1`. 

> > Note:" Aliasing via the `use` clause is currently not supported for enum variants, and will be added in later language versions

Reviewer 2: ```suggestion
> Note: Aliasing via the `use` clause is currently not supported for enum variants, and will be added in later language versions.
```

In selected contexts, the Move compiler infers the enum type from the context, and the qualification by the type name can be omitted:

```move
fun f(data: VersionedData) {
  match (data) { V1{..} => .., ..} // simple variant name OK
}
```

## Matching Enum Values 

The value of an enum value can be inspected using a match expression. For example:

```move
fun area(self: &Rectangle): u64 {
    match (self) {
        Circle{radius}           => mul_with_pi(*radius * *radius),
        Rectangle{width, height) => *width * *height
    }
}
```

Notice that above the value matched is an immutable reference to an enum value. We can also  consume values, or match over a mutable reference for interior updates:

```move

> fun scale_radius(self: &mut Rectanle, factor:  u64) {

Reviewer 2: ```suggestion
fun scale_radius(self: &mut Rectangle, factor:  u64) {
```

    match (self) {

>         Circle{radius} => *radius = *radius * factor,

Reviewer 2: Can we have an example (or reuse this example) to show that we can also do `Circle {radius: r}` in the match arm? This syntax may not be obvious to non-Rust folks (and it might not be obvious how else to avoid shadowing variables).

Author: Done

        _              => {} // do nothing if not a Circle
  }
}
```

The patterns provided in the match expression are evaluated sequentially, in order of textual experience, until a match is found. It is a compile time error if not all known patterns are covered.

Patterns can be nested and contain conditions, as in the following example:

```move
use Result::*;
let r : Result<Result<u64>> = Ok(Err(42));
let v = match (r)) {
  Ok(Err(c)) if c < 42  => 0,
  Ok(Err(c)) if c >= 42 => 1,

>   Ok(_)                 => 2,

Reviewer 2: I am confused by this suggestion. `Result` is not a variant?

>   Ok(_)                 => 2,

Reviewer: Sorry, I meant:
```
   Ok(Ok(_))     => 2,
```
Otherwise, the last match isn't needed to cover `Ok(Err(_))` as it's covered by the 3d clause.

>   Ok(_)                 => 2,

Reviewer: Note that you might want to not resolve a comment if there is confusion about it.  I had to hunt for this.

  _                     => 3 

> assert!(v == 1);

Reviewer 2: Missing ending `}`

```

Notice that in the above example, the last match clause covers both the patterns Ok(Err(\_)) as Err(\_). The earlier one is not covered since whether a conditional pattern matches is not decidable by the compiler; thus the first two clauses in the match expression above are not counting for match completeness.

Notice that while match completeness is enforced by the compiler, at  runtime it may not be guaranteed if code is upgraded. See the discussion of upgrade compatibility of enums here.

## Testing Enum Variants

With the `is` operator, one can examine whether a given enum value is of a given variant:

```move
let data: VersionedData;
if (data is VersionedData::V1) { .. }
```

The operator allows to specify a list of variants. Also, variants can be inferred if the type of the expression being tested is known:

```move
assert!(data is V1|V2);
```

## Selecting From Enum Values

It is possible to directly select a field from an enum value. Recall the definition of versioned data:

> ```

Reviewer 3: ```suggestion
```move
```

enum VersionedData has key {
  V1{name: String}
  V2{name: String, age: u64}
}
```

One can write code as below to directly select the fields of variants:

```move
let s: String;
let data1 = VersionedData::V1{name: s};
let data2 = VersionedData::V2{name: s, age: 20};
assert!(data1 is VersionData::V2);
assert!(data1.name == data2.name)
assert!(data2.age == 20); // data1.age will abort!
```

> Notice that field selection aborts if the enum value has no variant with the given field. This is the case for  `data1.age`. 

Author: Done

Field selection is only possible if the field is uniquely named and typed throughout all variants. Thus, the following yields in a compile time error:

> ```

Reviewer 3: ```suggestion
```move
```

enum VersionedData has key {
  V1{name: String}
  V3{name: u64}
}

data.name
 // ^^^^^ compile time error that `name` field selection is ambigious
```

## Using Enums Patterns in Lets

It is allowed to use an enum variant pattern in a let:

```move
let data: VersionData;

> let V1{name} = data;

Author: We could but its not that much needed as in Rust, as you can write:

```
if (!(x is V)) return;
let V{..} = x;
...
```

Simiilar for `if let`. But we should discuss.

```

Unpacking of the enum value will abort if the variant is not the expected one. If one wants to ensure that all variants of an enum are taken care of, a `match` expressions is recommended instead of a `let`. The `match` is checked at compile time whether all variants are covered. Otherwise, tools like the Move Prover can be used to verify that unexpected aborts cannot happen.

## Enum Type Upgrade Compatibility 

An enum type can be upgraded by another enum type if the new type only adds new variants at the end of the variant list. All variants already in the old enum type must be also, in the same order and starting from the beginning, in the new type. Consider the `VersionedData` type would have started as follows:

> ```

Reviewer 3: ```suggestion
```move
```

> An enum type can be upgraded by another enum type if the new type only adds new variants at the end of the variant list. All variants already in the old enum type must be also, in the same order and starting from the beginning, in the new type. Consider the `VersionedData` type would have started as follows:

Reviewer: ```suggestion
An enum type can be upgraded by another enum type if the new type only adds new variants at the end of the variant list. All variants present in the old enum type must also appear in the new type, in the same order and starting from the beginning. Consider the `VersionedData` type, which might have begun with a single version:
```

enum VersionedData has key {
  V1{name: String}
}
```

This type can be upgraded to the version we used so far in this text:

> ```

Reviewer 3: ```suggestion
```move
```

> This type can be upgraded to the version we used so far in this text:

Reviewer: ```suggestion
This type could be upgraded to the version we used so far in this text:
```

enum VersionedData has key {
  V1{name: String}
  V2{name: String, age: u64}
}
```

> Notice that the following upgrade would not be allowed, since the order of variants must be preserved:

Reviewer: ```suggestion
That following upgrade would not be allowed, since the order of variants must be preserved:
```
